### ISSUE TITLE
[Feature Request]Add SGDR, SGDW, AdamW and AdamWR

### ISSUE THREAD

**Issue by qianyizhang** [tensorflow/tensorflow#15237]
This is a duplicate request from pytorch issue, which I even reuse their issue title
https://github.com/pytorch/pytorch/issues/3790

And the fix from the paper seems to be trivial but I'm not sure how TF should approach this, perhaps just adding another parameter?

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by NProkoptsev** [comment]
Would be great if this added to Tensorflow.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by NProkoptsev** [comment]
Yes, this methods still haven't got implementation in tensorflow

**Comment by iron9light** [comment]
@qianyizhang @NProkoptsev SGDR was added as `tf.train.cosine_decay` and `tf.train.cosine_decay_restarts`
See https://github.com/tensorflow/tensorflow/blob/9bdb72e124e50e1b12b3286b38cbb1c971552741/tensorflow/python/training/learning_rate_decay.py#L487

**Comment by iron9light** [comment]
@qianyizhang @NProkoptsev For SGDW and AdamW, just check the code. https://github.com/tensorflow/tensorflow/blob/9bdb72e124e50e1b12b3286b38cbb1c971552741/tensorflow/core/kernels/training_ops.cc#L284 It seems they do not apply weight decay for Adam and Momentum.
I just checked the Apply Gradient part, and I think they won't apply weight decay when calculate gradient.
So I think nothing to FIX.
If you want something like AdamW (Adam + weight decay), you can add L2 regularization yourself.
I also checked the implementation of mxnet. They do support weight decay, and with the wrong way (based on that paper).
https://github.com/apache/incubator-mxnet/blob/d511d98aeadf4392f8c4c1289d15feecdd98bf81/src/operator/optimizer_op-inl.h#L718
Correct me if I'm wrong.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by NProkoptsev** [comment]
@iron9light in fact, author states that L2 regularization is not equal to weight decay. He proposes to move weight decay term to update of x_t.
The paper has updated recently, and it's more clear now.
https://arxiv.org/pdf/1711.05101.pdf

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tim9510019** [comment]
When will SGDW, AdamW be added to tensorflow??

**Comment by jonkoi** [comment]
Isn't SGDW just SGD with L2 reg?

**Comment by ebrevdo** [comment]
Seems so, I don't think there's anything actionable here.  Just add weight decay to the graph yourself.

**Comment by ebrevdo** [comment]
I misread.  I don't know anyone working on this; and it sounds like it would need to be implemented at the op level.  The interested implementer will need to write a custom kernel (+ GPU support if possible, + sparse update support) in tf.contrib.

**Comment by ebrevdo** [comment]
To clarify: the request is for implementations of AdamW, AdamWR from [this paper](https://arxiv.org/abs/1711.05101).

**Comment by facaiy** [comment]
@ebrevdo I think #17438 has implemented AdamW, AdamWR.

**Comment by PhilJd** [comment]
@facaiy 
I used the fact that the apply_ functions already get the precomputed gradient and Adam + Momentum optimizers don't compute values based on `var`, so it's equivalent to pre-decay. However, I just checked the apply functions of other optimizers and e.g. `Ftrl` computes factors based on `var`, so decoupling decay for such optimizers would indeed need a custom op. I'll leave a comment in the PR.

**Comment by JaeDukSeo** [comment]
So is this still an open issue? Please let me know if it is I would love to contribute. @ebrevdo

**Comment by JaeDukSeo** [comment]
Also I would love to implement AMSGrad (I did a tf implementation for fun here: http://bit.ly/2J6MRVb). And I was just wanting to know if this can be added as well.

**Comment by pierremac** [comment]
@JaeDukSeo that would be awesome! And as far as I know, it is still very open. I'm looking forward to trying it (but got no time to implement it myself).
Also, a few TF implementations of AMSGrad already exist. I've been using [this one](https://github.com/taki0112/AMSGrad-Tensorflow), which is significantly slower than "vanilla" Adam, but seems to be correct.

**Comment by PhilJd** [comment]
There's an existing PR for  weight decay  (linked above), but I was busy the last few weeks due to ICRA. My plan is to finish this next Sunday. However this doesn't include AMSgrad, but I guess that's a separate issue ;)

**Comment by phybrain** [comment]
how to use tf.train.cosine_decay_restarts to implement AdamWR?

**Comment by jonkoi** [comment]
@phybrain I can't have Adam optimizer update learning rate from cosine_decay_restarts. So I what I did is initializing a new optimizer at each training step with the new learning rate as parameter.
